src/realm/object-store/sync/async_open_task.cpp

```diff
@@ -170,4 +173,101 @@ void AsyncOpenTask::async_open_complete(AsyncOpenCallback&& callback,
 }
 
+void AsyncOpenTask::migrate_schema_or_complete(AsyncOpenCallback&& callback,
+                                               std::shared_ptr<_impl::RealmCoordinator> coordinator, Status status)
+{
+    util::CheckedUniqueLock lock(m_mutex);
+    if (!m_session)
+        return;
+    auto session = m_session;
+    lock.unlock();
+
+    auto pending_migration = [&] {
+        auto rt = coordinator->begin_read();
+        return _impl::sync_schema_migration::has_pending_migration(*rt);
+    }();
+
+    if (!pending_migration) {
+        wait_for_bootstrap_or_complete(std::move(callback), coordinator, status);
+        return;
+    }
+
+    // Sync schema migrations require setting a subscription initializer callback to bootstrap the data. The
+    // subscriptions in the current realm file may not be compatible with the new schema so cannot rely on them.
+    auto config = coordinator->get_config();
+    if (!config.sync_config->subscription_initializer) {
+        status = Status(ErrorCodes::SyncSchemaMigrationError,
+                        "Sync schema migrations must provide a subscription initializer callback in the sync config");
+        async_open_complete(std::move(callback), coordinator, status);
+        return;
+    }
+
+    // Migrate the schema.
+    //  * First upload the changes at the old schema version
+    //  * Then, delete the realm, reopen it, and bootstrap at new schema version
+    // The lifetime of the task is extended until bootstrap completes.
+    std::shared_ptr<AsyncOpenTask> self(shared_from_this());
+    session->wait_for_upload_completion([callback = std::move(callback), coordinator, session, self,
+                                         this](Status status) mutable {
+        {
+            util::CheckedLockGuard lock(m_mutex);
+            if (!m_session)
+                return; // Swallow all events if the task has been cancelled.
+        }
+
+        if (!status.is_ok()) {
+            self->async_open_complete(std::move(callback), coordinator, status);
+            return;
+        }
+
+        auto future = SyncSession::Internal::pause_async(*session);
+        // Wait until the SessionWrapper is done using the DBRef.
+        std::move(future).get_async([callback = std::move(callback), coordinator, self, this](Status status) mutable {
+            {
+                util::CheckedLockGuard lock(m_mutex);
+                if (!m_session)
+                    return; // Swallow all events if the task has been cancelled.
+            }
+
+            if (!status.is_ok()) {
+                self->async_open_complete(std::move(callback), coordinator, status);
+                return;
+            }
+
+            // Delete the realm file and reopen it.
+            try {
+                util::CheckedLockGuard lock(m_mutex);
+                auto config = coordinator->get_config();
+                m_session = nullptr;
+                coordinator->close();
```

Comment:
This is very unsafe. RealmCoordinator::close() is only safe to call if you can guarantee that no one else has a reference to the coordinator, which is not the case here. A simple example of something that'll break is if there are two async opens in progress at once.

Reply:
I think the assumption was always that for this to work only one realm coordinator/task is allowed to have the realm file opened (otherwise we cannot delete it)

Reply:
Well that isn't a very valid assumption.

Reply:
I see. I'll think of a different solution.

Reply:
it does already fail also on our ci https://github.com/realm/realm-core/issues/7325